### DISCUSSION TITLE
Can't set No Volume

### DISCUSSION THREAD

**Discussion by ceejayemm** [rern/rAudio#1594] [Fixed Issues]
I have just used image i20240108 to flash a new SD card on a Pi3B+ and then updated it to 20240210.  I configured this installation to use an IQaudIO DAC Pro (as I have done many times in the past).  As I play this server via my PC soundcard I tried to enable the 'No Volume' setting but this will not remain enabled.  In Player / Output I see only Bluetooth and Mixer Device (not IQaudIO DAC Pro as I would normally expect).  I do not have Bluetooth enabled on the System / On-board Devices and I do have IQaudIO DAC Pro selected in System / GPIO Devices / Audio I2S. I have rebooted the RPi several times to no avail.

Regards

Chris

**Comment by Vseason** [comment]
@rern Confirmed, and tested on a Pi4B 8Gb. I am unable to set anything other than “Mixer device” from Output > Volume Control, and the toggle for  Bit-Perfect > No Volume will not stay on.

Additionally, the selector for output devices seems to be missing.

**Reply by Vseason** [reply]
![IMG_6693](https://github.com/rern/rAudio/assets/107371732/18a0646c-f547-444c-8d6a-99ac597889fc)

**Comment by Vseason** [comment]
@rern 20240212 restores visibility of the output devices selector, but it is still not possible to set anything other than “Mixer device” from Output > Volume Control, and the toggle for Bit-Perfect > No Volume still will not stay on, regardless of the device selected.
![IMG_6699](https://github.com/rern/rAudio/assets/107371732/4835492b-34f0-4d8e-85c8-c0761b65dbcb)

**Comment by ceejayemm** [comment]
Been away for a few days so I have only just been able to get back to this. I have applied the latest UPDATE branch version (as at 17:15 GMT, Feb 14th).  Applying the UPDATE branch ended with:

Invalid card number 'scontents'.

and unfortunately with this version I am not able to access the Settings / Player page due to the following JSON error:

Errors: Unexpected token ',', ..."layname": , "name": "... is not valid JSON Copy
<{
  "page"  : "player"
, "login" : false
, "asoundcard"       : false
, "autoupdate"       : false
, "bluetooth"        : false
, "btaplayname"      : ""
, "buffer"           : false
, "bufferconf"       : { "KB": 4096 }
, "camilladsp"       : false
, "card"             : false
, "control"          : ""
, "counts"           : {
  "album": 338,
  "albumartist": 131,
  "albumyear": 609,
  "artist": 648,
  "composer": 1052,
  "conductor": 2,
  "dabradio": 0,
  "date": 163,
  "genre": 38,
  "latest": 0,
  "nas": 1,
  "playlists": 0,
  "sd": 0,
  "song": 5113,
  "usb": 0,
  "webradio": 4
}
, "crossfade"        : false
, "crossfadeconf"    : { "SEC": 0 }
, "custom"           : false
, "dabradio"         : false
, "device"           : { "aplayname": , "name": "IQaudIO DAC Pro", "card": false, "device": false, "hwmixer": false, "mixertype": "none" }
, "devicelist"       : {  "IQaudIO DAC HiFi pcm512x-hifi-0": "IQaudIO DAC HiFi pcm512x-hifi-0" }
, "devicewithbt"     : false
, "dop"              : false
, "equalizer"        : false
, "ffmpeg"           : true
, "lastupdate"       : "2024-02-11 <gr>• 10:05</gr>"
, "lists"            : {
  "albumignore" : false
, "mpdignore"   : true
, "nonutf8"     : false
}
, "mixerlist"        : false
, "normalization"    : false
, "novolume"         : true
, "outputbuffer"     : false
, "outputbufferconf" : { "KB": 8192 }
, "player"           : "mpd"
, "pllength"         : 9
, "replaygain"       : false
, "replaygainconf"   : {
  "TYPE"     : "album"
, "HARDWARE" : false
}
, "soxr"             : false
, "soxrconf"         : { "PLUGIN": "soxr", "QUALITY": "very high", "THREADS": 1 }
, "soxrcustomconf"   : { "PLUGIN": "soxr", "QUALITY": "custom", "PRECISION": 20, "PHASE_RESPONSE": 50, "PASSBAND_END": 91.3, "STOPBAND_BEGIN": 100, "ATTENUATION": 0, "FLAGS": 0 }
, "soxrquality"      : ""
, "state"            : "stop"
, "updatetime"       : "(MPD: 10m 34s + Cache: 14s)"
, "updating_db"      : false
, "version"          : "0.23.15-1"
, "volumempd"        : 1
}`
``

Chris

**Reply by Lubos-V** [reply]
Hi, just to make sure - #1603 seems to be related to the same issue (my bugreport) and also #1602 . I am just curious whether the title does the issue justice @ceejayemm .

**Comment by ceejayemm** [comment]
@rern

A reboot makes the situation worse, the Player/Output section disappears entirely. I now have a none working system. 

Regards

Chris

**Reply by ceejayemm** [reply]
@rern 

Further information:

This issue seems to have caused rAudio to have lost all knowledge of the installed DAC.  In an effort to force a re-installation of the DAC settings, I have changed the DAC setting several times in System / GPIO Devices / Audio - I2S. I have tried other IQaudIO DACs, various HifiBerry DACs and None with a reboot with every change, no difference noted. After each trial I have returned the setting to IQaudIO DAC Pro and still do not have a working system.  

The following information might help:

RAOffice:/srv/http/data/shm  #  cat /srv/http/data/shm/output
                                                 <--- Note blank line here
aplayname=""
name="IQaudIO DAC+"
card=
device=
hwmixer=
mixertype=software

RAOffice:/srv/http/data/shm  #  aplay -lL
ALSA lib conf.c:1245:(parse_value) card is not a string
ALSA lib conf.c:2012:(_snd_config_load_with_include) _toplevel_:2:18:Invalid argument
ALSA lib conf.c:4120:(config_file_open) /etc/asound.conf may be old or corrupted: consider to remove or fix it
ALSA lib conf.c:4042:(snd_config_hooks_call) function snd_config_hook_load returned error: Invalid argument
ALSA lib conf.c:4649:(snd_config_update_r) hooks failed, removing configuration
**** List of PLAYBACK Hardware Devices ****
ALSA lib conf.c:1245:(parse_value) card is not a string
ALSA lib conf.c:2012:(_snd_config_load_with_include) _toplevel_:2:18:Invalid argument
ALSA lib conf.c:4120:(config_file_open) /etc/asound.conf may be old or corrupted: consider to remove or fix it
ALSA lib conf.c:4042:(snd_config_hooks_call) function snd_config_hook_load returned error: Invalid argument
ALSA lib conf.c:4649:(snd_config_update_r) hooks failed, removing configuration
aplay: device_list:286: control open (0): Invalid argument


RAOffice:/srv/http/data/shm  #  cat /etc/asound.conf
defaults.pcm.card 
defaults.ctl.card


I hope this goes someway towards resolving this issue, please let me know if there is anything else I can provide.

Chris

**Reply by rern** [reply]
Hi Chris,

Try:
```sh
sed '/^defaults.*card *$/ s/$/ 0/' /etc/asound.conf
audioaplayname=$( < /srv/http/data/system/audio-aplayname )
card=$( aplay -l \
	| grep -m1 "$audioaplayname" \
	| sed -E 's/^card |: .*//g' )
sed "/^defaults.*card/ s/$/ $card/" /etc/asound.conf
/srv/http/bash/settings/player-conf.sh
```

**Reply by ceejayemm** [reply]
@rern 
Thanks for this code.  I have run it and this is what I get.  Is this what you expected:

RAOffice:~  #  sed '/^defaults.*card *$/ s/$/ 0/' /etc/asound.conf
audioaplayname=$( < $dirsystem/audio-aplayname )
card=$( aplay -l \
        | grep -m1 "$audioaplayname" \
        | sed -E 's/^card |: .*//g' )
sed "/^defaults.*card/ s/$/ $card/" /etc/asound.conf
/srv/http/bash/settings/player-conf.sh
defaults.pcm.card  0
defaults.ctl.card  0

ALSA lib conf.c:1245:(parse_value) card is not a string
ALSA lib conf.c:2012:(_snd_config_load_with_include) _toplevel_:2:18:Invalid argument
ALSA lib conf.c:4120:(config_file_open) /etc/asound.conf may be old or corrupted: consider to remove or fix it
ALSA lib conf.c:4042:(snd_config_hooks_call) function snd_config_hook_load returned error: Invalid argument
ALSA lib conf.c:4649:(snd_config_update_r) hooks failed, removing configuration
aplay: device_list:286: control open (0): Invalid argument
defaults.pcm.card  
defaults.ctl.card  

grep: /srv/http/data/shm/output: No such file or directory
grep: /srv/http/data/shm/output: No such file or directory
grep: /srv/http/data/shm/output: No such file or directory
grep: /srv/http/data/shm/output: No such file or directory
RAOffice:~  #  

Thanks

Chris

**Comment by ceejayemm** [comment]
@rern 

I see that you have released Update 20230218 where some further work on the 'No Volume' setting issue has been done. However this release is not  currently available to me for some reason so I have applied the latest UPDATE branch (which should be equivalent  to the MAIN release I guess). I still do not have any output details, so not a working system, as detailed earlier in this issue..

Chris

**Comment by rern** [comment]
Please try today's update, 20240219.

**Reply by Lubos-V** [reply]
on a first sight it looks good @rern - update went fine, all parts of the UI are accessible. will do some more thorough testing later and will report if i find something peculiar.

**Comment by ceejayemm** [comment]
@rern 
Applied update 20240219 as requested, Player / Output screen is restored (sort of - see below) and JSON error seems to have gone. The new Player / Output screen shows the following:

![rAudio-20240219](https://github.com/rern/rAudio/assets/25418025/7292395c-759f-4470-a520-95d26831620d)

This is with (on System screen) Bluetooth turned on and IQaudIO DAC Pro selected.  Is this what you would expect ?

I still cannot produce any audio output, the Player / Music Player Daemon log snippet shows:

mpd 0.23.15-1
# /srv/http/data/mpdconf/mpd.conf
bind_to_address     "0.0.0.0"
db_file             "/srv/http/data/mpd/mpd.db"
music_directory     "/mnt/MPD"
playlist_directory  "/srv/http/data/playlists"
state_file          "/var/lib/mpd/mpdstate"
user                "mpd"
input {
	plugin  "curl"
}
decoder {
	plugin   "ffmpeg"
	enabled  "yes"
}
audio_output {
	name           ""
	device         "hw:,"
	type           "alsa"
	auto_resample  "no"
	auto_format    "no"
	mixer_type     "none"
}

mpd.service- Music Player Daemon
     Loaded: loaded (/usr/lib/systemd/system/mpd.service; disabled; preset: disabled)
    Drop-In: /usr/lib/systemd/system/mpd.service.d
             └─00-arch.conf
             /etc/systemd/system/mpd.service.d
             └─override.conf
     Active: active (running) since Mon 2024-02-19 16:57:10 UTC; 10min ago
       Docs: man:mpd(1)
             man:mpd.conf(5)
   Main PID: 541 (mpd)
      Tasks: 7 (limit: 926)
        CPU: 2.609s
     CGroup: /system.slice/mpd.service
             └─541 /usr/bin/mpd --systemd /srv/http/data/mpdconf/mpd.conf

Feb 19 16:59:20 RAOffice mpd[541]: exception: Failed to open "" (alsa); Failed to open ALSA device "hw:,": Invalid argument
Feb 19 16:59:20 RAOffice mpd[541]: exception: Failed to open "" (alsa); Failed to open ALSA device "hw:,": Invalid argument
Feb 19 16:59:20 RAOffice mpd[541]: player: problems opening audio device while playing "NAS/NAS_Music/Rick Wakeman/The Six Wives Of Henry VIII Live/1-01 Tudorture   1485.m4a"
Feb 19 16:59:25 RAOffice mpd[541]: ALSA lib conf.c:1245:(parse_value) card is not a string
Feb 19 16:59:25 RAOffice mpd[541]: ALSA lib conf.c:2012:(_snd_config_load_with_include) _toplevel_:2:18:Invalid argument
Feb 19 16:59:25 RAOffice mpd[541]: ALSA lib conf.c:4120:(config_file_open) /etc/asound.conf may be old or corrupted: consider to remove or fix it
Feb 19 16:59:25 RAOffice mpd[541]: ALSA lib conf.c:4042:(snd_config_hooks_call) function snd_config_hook_load returned error: Invalid argument
Feb 19 16:59:25 RAOffice mpd[541]: ALSA lib conf.c:4649:(snd_config_update_r) hooks failed, removing configuration
Feb 19 16:59:25 RAOffice mpd[541]: exception: Failed to open "" (alsa); Failed to open ALSA device "hw:,": Invalid argument
Feb 19 16:59:25 RAOffice mpd[541]: exception: Failed to open "" (alsa); Failed to open ALSA device "hw:,": Invalid argument

So it still looks like the /etc/asound.conf file is corrupted or incorrectly formatted.  This file has the following information:

RAOffice:~  #  cat /etc/asound.conf
defaults.pcm.card 
defaults.ctl.card 

RAOffice:~  #  

and 

RAOffice:~  #  amixer scontents
ALSA lib conf.c:1245:(parse_value) card is not a string
ALSA lib conf.c:2012:(_snd_config_load_with_include) _toplevel_:2:18:Invalid argument
ALSA lib conf.c:4120:(config_file_open) /etc/asound.conf may be old or corrupted: consider to remove or fix it
ALSA lib conf.c:4042:(snd_config_hooks_call) function snd_config_hook_load returned error: Invalid argument
ALSA lib conf.c:4649:(snd_config_update_r) hooks failed, removing configuration
amixer: Mixer attach default error: Invalid argument
RAOffice:~  #  

and

RAOffice:~  #  cat /srv/http/data/shm/output
mixer=false
aplayname=""
name=""
card=
device=
mixertype=none
RAOffice:~  #  

Where do we go from here ?

Regards.

Chris

**Reply by ceejayemm** [reply]
Further information.

Removing the I2S device (IQaudIO DAC Pro) and setting this value to 'None' then disabling the No Volume setting (Player / Bit Perfect) allows the Player / Output options to be displayed and I can play audio, albeit via the onboard headphone socket. It would appear to be a problem relating to the setup / configuration of the I2S sound card.

Chris

**Reply by clriis** [reply]
@ceejayemm 

Exactly the same issue with I2S 'Allo Boss DAC' on Raspberry Pi 3B Rev 1.2 and rAudio 20240219

**Comment by ceejayemm** [comment]
I have kept this update to my own issue for reasons that will become clear below.

As requested in #1605, as I already have 20240219 already installed

RAOffice:~  #  cat /etc/asound.conf
                              <--- Blank line
RAOffice:~  #

RAOffice:~  #  aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: b1 [bcm2835 HDMI 1], device 0: bcm2835 HDMI 1 [bcm2835 HDMI 1]
  Subdevices: 4/4
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
card 0: b1 [bcm2835 HDMI 1], device 1: bcm2835 HDMI 1 [bcm2835 HDMI 1]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: Headphones [bcm2835 Headphones], device 0: bcm2835 Headphones [bcm2835 Headphones]
  Subdevices: 4/4
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
card 2: IQaudIODAC [IQaudIODAC], device 0: IQaudIO DAC HiFi pcm512x-hifi-0 [IQaudIO DAC HiFi pcm512x-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
RAOffice:~  #  

After requested update 
RAOffice:~  #  cat /etc/asound.conf
defaults.pcm.card 0
defaults.ctl.card 0

RAOffice:~  #  

Player settings: Device empty

Reapplied request update to /etc/asound.conf and rebooted, this time:

RAOffice:~  #  cat /etc/asound.conf
defaults.pcm.card 2
defaults.ctl.card 2

RAOffice:~  #

RAOffice:~  #  aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: b1 [bcm2835 HDMI 1], device 0: bcm2835 HDMI 1 [bcm2835 HDMI 1]
  Subdevices: 4/4
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
card 0: b1 [bcm2835 HDMI 1], device 1: bcm2835 HDMI 1 [bcm2835 HDMI 1]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: Headphones [bcm2835 Headphones], device 0: bcm2835 Headphones [bcm2835 Headphones]
  Subdevices: 4/4
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
card 2: IQaudIODAC [IQaudIODAC], device 0: IQaudIO DAC HiFi pcm512x-hifi-0 [IQaudIO DAC HiFi pcm512x-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0

Output Device settings now in place and I can make adjustments as needed. I have audio output via the IQaudIO DAC Pro as well!

I have no idea why this worked after the second update/reboot. For the time being, on the age old basis "If it ain't broke, don't fix it", I  am now leaving this alone :-)

Chris

**Comment by ceejayemm** [comment]
@rern 
I can confirm that with today's update things seem to have got back to normal.

Many thanks

Chris
